composer/algorithms/factorize/factorize_modules.py

```diff
@@ -0,0 +1,441 @@
+        if not self.should_factorize(self.latent_channels):
+            raise ValueError(
+                f"latent_channels {self.latent_size} is not small enough to merit factorization! Must be <= {self._max_rank_with_speedup()}"
+            )
+
+        # this one produces identical output as a regular Conv2d would,
+        # except with fewer output channels
+        conv0 = nn.Conv2d(self.in_channels,
+                          self.latent_channels,
+                          self.kernel_size,
+                          bias=False,
+                          **self.convolution_kwargs)
+        # this one increases the number of output channels
+        conv1 = nn.Conv2d(self.latent_channels, self.out_channels, kernel_size=1, bias=True)
+        return conv0, conv1
+
+    # wrap shared fields in read-only properties matching the torch conv module API
+    @property
+    def in_channels(self) -> int:
+        """See :class:`torch.nn.Conv2d`."""
+        return self.in_size
+
+    @property
+    def out_channels(self) -> int:
+        """See :class:`torch.nn.Conv2d`."""
+        return self.out_size
+
+    @property
+    def latent_channels(self) -> int:
+        """The number of of output channels for the first convolution, which is also the number of input channels for the second convolution"""
+        return self.latent_size
+
+    def solution_for_rank(self, input: torch.Tensor, rank: int) -> LowRankSolution:
+        weight0 = self.module0.weight
+        bias0 = self.module0.bias
+        weight1, bias1 = self.module1.weight, self.module1.bias
+
+        assert (bias0 is None) or isinstance(bias0, torch.Tensor)
+        assert isinstance(bias1, torch.Tensor)
+        assert isinstance(weight0, torch.Tensor)
+        assert isinstance(weight1, torch.Tensor)
+
+        return factorize_conv2d(input, weight0, weight1, rank=rank, biasA=bias0, biasB=bias1, **self.convolution_kwargs)
+
+    def apply_solution(self, solution: LowRankSolution):
+        self.latent_size = solution.rank
+        self.module0.out_channels = solution.rank
+        self.module1.in_channels = solution.rank
+        _apply_solution_to_module_parameters(solution, self.module0, self.module1, transpose=False)
+
+    @staticmethod
+    def max_allowed_latent_features(in_features: int, out_features: int, kernel_size: _size_2_t) -> int:
+        """Returns the largest latent channel count that reduces the number of multiply-adds
+
+        Args:
+            in_channels: number of channels in the input image
+            out_channels: number of channels produced by the convolution
+            kernel_size: size of the convolving kernel
+
+        Returns:
+            latent_channels: the largest allowable number of latent channels
+        """
+        return max_rank_with_possible_speedup(in_features, out_features, kernel_size=kernel_size)
+
+    @staticmethod
+    def from_conv2d(module: torch.nn.Conv2d, module_ix: int = -1, **kwargs) -> 'FactorizedConv2d':
```

Comment:
At the top of the file, you can add a `from __future__ import annotations`. Then, you don't need to use quotes for typing annotations.
```suggestion
    def from_conv2d(module: torch.nn.Conv2d, module_ix: int = -1, **kwargs) -> FactorizedConv2d:
```